### PR TITLE
feat: Consider Holiday List in Student Leave Application and Attendance

### PR DESCRIPTION
**Pull request by ruchamahabal**
Documentation PR: https://github.com/frappe/erpnext_documentation/pull/183

- Validations for Holiday List in Student Attendance

   attendance_validation

- **Total Leave Days** field added in Student Leave Application which is = days between `from_date` and `to_date` - number of holidays

   total_leave_days

- Student Attendance record creation on Leave Application submission skipped for holidays
- Holidays are shown as "H" in the Student Monthly Attendance Sheet

   monthly

- Fixed other attendance related reports
   - Student Batch wise Attendance
   - Absent Student Report
      
attendance_validation

- Added Test

### CODE REVIEW

**Review by Anurag810 — COMMENTED** [review]
@ruchamahabal I think we can use generalized this with our attendance, monthly attendance sheet, leave application in HR. both looks same. it was just a thought. As it sales order for healthcare.

File: erpnext/education/doctype/student_attendance/student_attendance.py [hunk 1]
```diff
@@ -66,3 +68,17 @@ def validate_duplication(self):
 			record = get_link_to_form('Attendance Record', attendance_record)
 			frappe.throw(_('Student Attendance record {0} already exists against the Student {1}')
 				.format(record, frappe.bold(self.student)), title=_('Duplicate Entry'))
+
+	def validate_is_holiday(self):
+		holiday_list = get_holiday_list()
+		if is_holiday(holiday_list, self.date):
+			frappe.throw(_('Attendance cannot be marked for {0} as it is a holiday.').format(frappe.bold(self.date)))
```
Comment (nabinhait) [inline erpnext/education/doctype/student_attendance/student_attendance.py]: Date should be displayed in user format

File: erpnext/education/report/absent_student_report/absent_student_report.py [hunk 2]
```diff
@@ -15,6 +17,11 @@ def execute(filters=None):
 	columns = get_columns(filters)
 	date = filters.get("date")
 
+	holiday_list = get_holiday_list()
+	if is_holiday(holiday_list, filters.get("date")):
+		msgprint(_("No attendance has been marked for {0} as it is a Holiday").format(frappe.bold(filters.get("date"))))
```
Comment (nabinhait) [inline erpnext/education/report/absent_student_report/absent_student_report.py]: Date in user format

File: erpnext/education/report/student_batch_wise_attendance/student_batch_wise_attendance.py [hunk 3]
```diff
@@ -5,13 +5,19 @@
...
 
 def execute(filters=None):
 	if not filters: filters = {}
 
 	if not filters.get("date"):
 		msgprint(_("Please select date"), raise_exception=1)
 
+	holiday_list = get_holiday_list()
+	if is_holiday(holiday_list, filters.get("date")):
+		msgprint(_("No attendance has been marked for {0} as it is a Holiday").format(frappe.bold(filters.get("date"))))
```
Comment (nabinhait) [inline erpnext/education/report/student_batch_wise_attendance/student_batch_wise_attendance.py]: Date in user format

**Review by nabinhait — CHANGES_REQUESTED** [review]

**Review by nabinhait — APPROVED** [review]

**Pull request merged**